File: bindings/csharp/SpecUtils/Measurement.cs
using System.Runtime.InteropServices;

namespace SpecUtils;

/// <summary>
/// Represents a single spectrum measurement/record within a SpecFile.
/// May be owned (caller must dispose) or borrowed (lifetime tied to parent SpecFile).
/// </summary>
public class Measurement : IDisposable
{
    internal IntPtr Handle;
    private bool _ownsHandle;
    private bool _disposed;

    // Hold a reference to parent SpecFile to prevent GC when we have borrowed pointers
    private readonly SpecFile? _parent;

    /// <summary>Creates a new empty owned Measurement.</summary>
    public Measurement()
    {
        Handle = NativeMethods.SpecUtils_Measurement_create();
        _ownsHandle = true;
    }

    internal Measurement(IntPtr handle, bool ownsHandle, SpecFile? parent = null)
    {
        Handle = handle;
        _ownsHandle = ownsHandle;
        _parent = parent;
    }

    /// <summary>Creates a deep copy (owned by caller).</summary>
    public Measurement Clone()
    {
        ThrowIfDisposed();
        IntPtr clone = NativeMethods.SpecUtils_Measurement_clone(Handle);
        return new Measurement(clone, ownsHandle: true);
    }

    /// <summary>Resets to initial empty state. Only valid on owned measurements.</summary>
    public void Reset()
    {
        ThrowIfDisposed();
        NativeMethods.SpecUtils_Measurement_reset(Handle);
    }

    /// <summary>Called when ownership is transferred to a SpecFile.</summary>
    internal void TransferOwnership()
    {
        _ownsHandle = false;
    }

    // ---- Properties ----

    public string Description
    {
        get { ThrowIfDisposed(); return Marshal.PtrToStringUTF8(NativeMethods.SpecUtils_Measurement_description(Handle)) ?? ""; }
        set { ThrowIfDisposed(); NativeMethods.SpecUtils_Measurement_set_description(Handle, value); }
    }

    public string SourceString
    {
        get { ThrowIfDisposed(); return Marshal.PtrToStringUTF8(NativeMethods.SpecUtils_Measurement_source_string(Handle)) ?? ""; }
        set { ThrowIfDisposed(); NativeMethods.SpecUtils_Measurement_set_source_string(Handle, value); }
    }

    public string Title
    {
        get { ThrowIfDisposed(); return Marshal.PtrToStringUTF8(NativeMethods.SpecUtils_Measurement_title(Handle)) ?? ""; }
        set { ThrowIfDisposed(); NativeMethods.SpecUtils_Measurement_set_title(Handle, value); }
    }

    public long StartTimeMicroseconds
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_start_time_usecs(Handle); }
        set { ThrowIfDisposed(); NativeMethods.SpecUtils_Measurement_set_start_time_usecs(Handle, value); }
    }

    public char PcfTag
    {
        get { ThrowIfDisposed(); return (char)NativeMethods.SpecUtils_Measurement_pcf_tag(Handle); }
        set { ThrowIfDisposed(); NativeMethods.SpecUtils_Measurement_set_pcf_tag(Handle, (byte)value); }
    }

    public uint NumberGammaChannels
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_number_gamma_channels(Handle); }
    }

    public float RealTime
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_real_time(Handle); }
    }

    public float LiveTime
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_live_time(Handle); }
    }

    public float NeutronLiveTime
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_neutron_live_time(Handle); }
    }

    public double GammaCountSum
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_gamma_count_sum(Handle); }
    }

    public double NeutronCountSum
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_neutron_count_sum(Handle); }
    }

    public bool IsOccupied
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_is_occupied(Handle); }
    }

    public bool ContainedNeutron
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_contained_neutron(Handle); }
    }

    public int SampleNumber
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_sample_number(Handle); }
        set { ThrowIfDisposed(); NativeMethods.SpecUtils_Measurement_set_sample_number(Handle, value); }
    }

    public string DetectorName
    {
        get { ThrowIfDisposed(); return Marshal.PtrToStringUTF8(NativeMethods.SpecUtils_Measurement_detector_name(Handle)) ?? ""; }
        set { ThrowIfDisposed(); NativeMethods.SpecUtils_Measurement_set_detector_name(Handle, value); }
    }

    public float Speed
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_speed(Handle); }
    }

    public OccupancyStatus OccupancyStatus
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_occupancy_status(Handle); }
        set { ThrowIfDisposed(); NativeMethods.SpecUtils_Measurement_set_occupancy_status(Handle, value); }
    }

    public bool HasGpsInfo
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_has_gps_info(Handle); }
    }

    public double Latitude
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_latitude(Handle); }
    }

    public double Longitude
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_longitude(Handle); }
    }

    public long PositionTimeMicroseconds
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_position_time_microsec(Handle); }
    }

    public float DoseRate
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_dose_rate(Handle); }
    }

    public float ExposureRate
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_exposure_rate(Handle); }
    }

    public string DetectorTypeName
    {
        get { ThrowIfDisposed(); return Marshal.PtrToStringUTF8(NativeMethods.SpecUtils_Measurement_detector_type(Handle)) ?? ""; }
    }

    public QualityStatus QualityStatus
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_quality_status(Handle); }
    }

    public SourceType SourceType
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_source_type(Handle); }
        set { ThrowIfDisposed(); NativeMethods.SpecUtils_Measurement_set_source_type(Handle, value); }
    }

    public uint MemorySize
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_memmorysize(Handle); }
    }

    public uint DerivedDataProperties
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_derived_data_properties(Handle); }
    }

    // ---- Remarks and warnings ----

    public uint NumberRemarks
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_number_remarks(Handle); }
    }

    public string GetRemark(int index)
    {
        ThrowIfDisposed();
        return Marshal.PtrToStringUTF8(
            NativeMethods.SpecUtils_Measurement_remark(Handle, (uint)index)) ?? "";
    }

    public uint NumberParseWarnings
    {
        get { ThrowIfDisposed(); return NativeMethods.SpecUtils_Measurement_number_parse_warnings(Handle); }
    }

    public string GetParseWarning(int index)
    {
        ThrowIfDisposed();
        return Marshal.PtrToStringUTF8(
            NativeMethods.SpecUtils_Measurement_parse_warning(Handle, (uint)index)) ?? "";
    }

    // ---- Data access ----

    /// <summary>
    /// Returns a copy of the gamma channel counts as a float array.
    /// Returns null if no gamma data is present.
    /// </summary>
    public float[]? GetGammaChannelCounts()
    {
        ThrowIfDisposed();
        uint nChannels = NativeMethods.SpecUtils_Measurement_number_gamma_channels(Handle);
        if (nChannels == 0)
            return null;

        IntPtr ptr = NativeMethods.SpecUtils_Measurement_gamma_channel_counts(Handle);
        if (ptr == IntPtr.Zero)
            return null;

        float[] counts = new float[nChannels];
        Marshal.Copy(ptr, counts, 0, (int)nChannels);
        return counts;
    }

    /// <summary>
    /// Returns a copy of the energy bounds (lower channel energies) as a float array.
    /// The array has NumberGammaChannels + 1 entries.
    /// Returns null if no energy calibration is set.
    /// </summary>
    public float[]? GetEnergyBounds()
    {
        ThrowIfDisposed();
        uint nChannels = NativeMethods.SpecUtils_Measurement_number_gamma_channels(Handle);
        if (nChannels == 0)
            return null;

        IntPtr ptr = NativeMethods.SpecUtils_Measurement_energy_bounds(Handle);
        if (ptr == IntPtr.Zero)
            return null;

        float[] energies = new float[nChannels + 1];
        Marshal.Copy(ptr, energies, 0, (int)(nChannels + 1));
        return energies;
    }

    /// <summary>
    /// Gets the energy calibration for this measurement (borrowed, do not dispose).
    /// </summary>
    public EnergyCalibration? GetEnergyCalibration()
    {
        ThrowIfDisposed();
        IntPtr ptr = NativeMethods.SpecUtils_Measurement_energy_calibration(Handle);
        return ptr == IntPtr.Zero ? null : new EnergyCalibration(ptr, ownsHandle: false);
    }

    /// <summary>
    /// Returns the gamma integral between the specified energies.
    /// </summary>
    public double GammaIntegral(float lowerEnergy, float upperEnergy)
    {
        ThrowIfDisposed();
        return NativeMethods.SpecUtils_Measurement_gamma_integral(Handle, lowerEnergy, upperEnergy);
    }

    /// <summary>
    /// Returns the sum of gamma counts between the specified bin indices.
    /// </summary>
    public double GammaChannelsSum(uint startBin, uint endBin)
    {
        ThrowIfDisposed();
        return NativeMethods.SpecUtils_Measurement_gamma_channels_sum(Handle, startBin, endBin);
    }

    // ---- Setters ----

    /// <summary>Sets gamma channel counts, live time, and real time.</summary>
    public void SetGammaCounts(float[] counts, float liveTime, float realTime)
    {
        ThrowIfDisposed();
        NativeMethods.SpecUtils_Measurement_set_gamma_counts(
            Handle, counts, (uint)counts.Length, liveTime, realTime);
    }

    /// <summary>Sets neutron counts from one or more tubes.</summary>
    public void SetNeutronCounts(float[] counts, float neutronLiveTime)
    {
        ThrowIfDisposed();
        NativeMethods.SpecUtils_Measurement_set_neutron_counts(
            Handle, counts, (uint)counts.Length, neutronLiveTime);
    }

    public bool SetStartTimeString(string startTimeStr)
    {
        ThrowIfDisposed();
        return NativeMethods.SpecUtils_Measurement_set_start_time_str(Handle, startTimeStr);
    }

    public void SetPosition(double longitude, double latitude, long positionTimeMicrosec = 0)
    {
        ThrowIfDisposed();
        NativeMethods.SpecUtils_Measurement_set_position(Handle, longitude, latitude, positionTimeMicrosec);
    }

    public bool CombineGammaChannels(uint nchannels)
    {
        ThrowIfDisposed();
        return NativeMethods.SpecUtils_Measurement_combine_gamma_channels(Handle, nchannels);
    }

    public bool Rebin(CountedRefEnergyCalibration cal)
    {
        ThrowIfDisposed();
        return NativeMethods.SpecUtils_Measurement_rebin(Handle, cal.Handle);
    }

    public bool SetEnergyCalibration(CountedRefEnergyCalibration cal)
    {
        ThrowIfDisposed();
        return NativeMethods.SpecUtils_Measurement_set_energy_calibration(Handle, cal.Handle);
    }

    // ---- IDisposable ----

    private void ThrowIfDisposed()
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
    }

    protected virtual void Dispose(bool disposing)
    {
        if (!_disposed)
        {
            if (_ownsHandle && Handle != IntPtr.Zero)
            {
                NativeMethods.SpecUtils_Measurement_destroy(Handle);
            }
            Handle = IntPtr.Zero;
            _disposed = true;
        }
    }

    ~Measurement()
    {
        Dispose(false);
    }

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }
}

File: bindings/java/src/main/java/gov/sandia/specutils/Measurement.java
package gov.sandia.specutils;

import gov.sandia.specutils.internal.HandleCleaner;
import gov.sandia.specutils.internal.HandleCleaner.Registration;
import gov.sandia.specutils.internal.Native;

/**
 * A single spectrum record (one detector, one sample interval) within a
 * {@link SpecFile}.
 *
 * <p>Instances may be <i>owned</i> (created via {@link #Measurement()} or
 * via {@link SpecFile#sumMeasurements}) or obtained from a SpecFile via
 * {@link SpecFile#measurement(int)}. The latter are <i>reference-counted</i>:
 * they keep the underlying measurement alive independent of the SpecFile, so
 * they remain valid even after the SpecFile is modified or closed. Either kind
 * should be {@link #close()}d when done. Transferring ownership to a SpecFile
 * (via {@link SpecFile#addMeasurement}) flips an owned measurement to
 * non-owning on success.
 */
public final class Measurement implements AutoCloseable {

    static { Native.init(); }

    long handle;
    private boolean ownsHandle;
    // For measurements obtained from a SpecFile: an owning reference-counted handle that keeps the
    // underlying measurement alive (0 for owned measurements).  `handle` is then a view into it.
    private long refHandle;
    // Held to keep the parent SpecFile from being GC'd while in use (not required for lifetime
    // safety any more - the ref handle owns the measurement - but harmless and convenient).
    @SuppressWarnings("unused")
    private final SpecFile parent;
    private Registration cleanerRegistration;
    private final Object cleanerOwner = new Object();

    /** Creates a new owned, empty Measurement. */
    public Measurement() {
        this.handle = Native.measurementCreate();
        if (this.handle == 0L) {
            throw new SpecUtilsException("measurementCreate returned null");
        }
        this.ownsHandle = true;
        this.parent = null;
        this.cleanerRegistration = registerCleaner();
    }

    /** Internal: wraps an existing owned handle (e.g. from clone or sum). */
    Measurement(long handle, boolean ownsHandle, SpecFile parent) {
        this.handle = handle;
        this.ownsHandle = ownsHandle;
        this.refHandle = 0L;
        this.parent = parent;
        this.cleanerRegistration = ownsHandle ? registerCleaner() : null;
    }

    /** Internal: wraps a reference-counted handle obtained from a SpecFile. */
    private Measurement(long viewHandle, long refHandle, SpecFile parent) {
        this.handle = viewHandle;
        this.ownsHandle = false;
        this.refHandle = refHandle;
        this.parent = parent;
        this.cleanerRegistration = registerCleaner();
    }

    /**
     * Wraps a reference-counted measurement handle (from Native.specFileGetMeasurementRef*).
     * The ref keeps the underlying measurement alive, so this wrapper stays valid even after the
     * parent SpecFile is modified or closed.
     */
    static Measurement fromRef(long refHandle, SpecFile parent) {
        final long view = Native.measurementPtrFromRef(refHandle);
        if (view == 0L) {
            Native.countedRefMeasurementDestroy(refHandle);
            throw new SpecUtilsException("measurementPtrFromRef returned null");
        }
        return new Measurement(view, refHandle, parent);
    }

    private Registration registerCleaner() {
        // Capture what this wrapper is responsible for freeing: a reference-counted handle (for
        // measurements from a SpecFile) or, for owned measurements, the measurement itself.
        final long ref = this.refHandle;
        final long ownedHandle = this.ownsHandle ? this.handle : 0L;
        return HandleCleaner.register(cleanerOwner, () -> {
            if (ref != 0L) {
                Native.countedRefMeasurementDestroy(ref);
            } else if (ownedHandle != 0L) {
                Native.measurementDestroy(ownedHandle);
            }
        });
    }

    /**
     * Called by SpecFile.addMeasurement when ownership has been transferred
     * to the parent. Makes {@link #close()} a no-op.
     */
    synchronized void transferOwnership() {
        this.ownsHandle = false;
        if (cleanerRegistration != null) {
            cleanerRegistration.deregister();
            cleanerRegistration = null;
        }
    }

    /** Creates a deep copy, owned by the caller. */
    public Measurement clone_() {
        throwIfClosed();
        long h = Native.measurementClone(handle);
        if (h == 0L) {
            throw new SpecUtilsException("measurementClone returned null");
        }
        return new Measurement(h, true, null);
    }

    public void reset() {
        throwIfClosed();
        Native.measurementReset(handle);
    }

    public int memorySize() {
        throwIfClosed();
        return Native.measurementMemorySize(handle);
    }

    // ----- Metadata getters/setters -----

    public String description()         { throwIfClosed(); return Native.measurementDescription(handle); }
    public void setDescription(String s) { throwIfClosed(); Native.measurementSetDescription(handle, s); }

    public String sourceString()         { throwIfClosed(); return Native.measurementSourceString(handle); }
    public void setSourceString(String s) { throwIfClosed(); Native.measurementSetSourceString(handle, s); }

    public String title()         { throwIfClosed(); return Native.measurementTitle(handle); }
    public void setTitle(String s) { throwIfClosed(); Native.measurementSetTitle(handle, s); }

    public String detectorName()         { throwIfClosed(); return Native.measurementDetectorName(handle); }
    public void setDetectorName(String s) { throwIfClosed(); Native.measurementSetDetectorName(handle, s); }

    public String detectorType() { throwIfClosed(); return Native.measurementDetectorType(handle); }

    public long startTimeMicroseconds() {
        throwIfClosed();
        return Native.measurementStartTimeUsecs(handle);
    }

    public void setStartTimeMicroseconds(long microsecondsSinceEpoch) {
        throwIfClosed();
        Native.measurementSetStartTimeUsecs(handle, microsecondsSinceEpoch);
    }

    /** Throws {@link SpecUtilsException} if the string cannot be parsed. */
    public void setStartTimeString(String startTime) {
        throwIfClosed();
        if (!Native.measurementSetStartTimeStr(handle, startTime)) {
            throw new SpecUtilsException("setStartTimeString: could not parse '"
                + startTime + "'");
        }
    }

    public char pcfTag()        { throwIfClosed(); return Native.measurementPcfTag(handle); }
    public void setPcfTag(char c) { throwIfClosed(); Native.measurementSetPcfTag(handle, c); }

    public int sampleNumber()            { throwIfClosed(); return Native.measurementSampleNumber(handle); }
    public void setSampleNumber(int n)   { throwIfClosed(); Native.measurementSetSampleNumber(handle, n); }

    // ----- Timing & counts -----

    public float realTime()         { throwIfClosed(); return Native.measurementRealTime(handle); }
    public float liveTime()         { throwIfClosed(); return Native.measurementLiveTime(handle); }
    public float neutronLiveTime()  { throwIfClosed(); return Native.measurementNeutronLiveTime(handle); }
    public double gammaCountSum()   { throwIfClosed(); return Native.measurementGammaCountSum(handle); }
    public double neutronCountSum() { throwIfClosed(); return Native.measurementNeutronCountSum(handle); }

    public boolean isOccupied()        { throwIfClosed(); return Native.measurementIsOccupied(handle); }
    public boolean containedNeutron()  { throwIfClosed(); return Native.measurementContainedNeutron(handle); }

    public int numberGammaChannels()   { throwIfClosed(); return Native.measurementNumberGammaChannels(handle); }

    /** @return copy of gamma channel counts; empty array if unset. */
    public float[] gammaChannelCounts() {
        throwIfClosed();
        return Native.measurementGammaChannelCounts(handle);
    }

    /** @return copy of lower channel energies (length = numChannels + 1), empty if unset. */
    public float[] energyBounds() {
        throwIfClosed();
        return Native.measurementEnergyBounds(handle);
    }

    public double gammaIntegral(float lowerEnergy, float upperEnergy) {
        throwIfClosed();
        return Native.measurementGammaIntegral(handle, lowerEnergy, upperEnergy);
    }

    public double gammaChannelsSum(int startBin, int endBin) {
        throwIfClosed();
        return Native.measurementGammaChannelsSum(handle, startBin, endBin);
    }

    public int derivedDataProperties() {
        throwIfClosed();
        return Native.measurementDerivedDataProperties(handle);
    }

    public void setGammaCounts(float[] counts, float liveTime, float realTime) {
        throwIfClosed();
        Native.measurementSetGammaCounts(handle, counts, liveTime, realTime);
    }

    public void setNeutronCounts(float[] counts, float neutronLiveTime) {
        throwIfClosed();
        Native.measurementSetNeutronCounts(handle, counts, neutronLiveTime);
    }

    public boolean combineGammaChannels(int nChannel) {
        throwIfClosed();
        return Native.measurementCombineGammaChannels(handle, nChannel);
    }

    public void rebin(EnergyCalibration cal) {
        throwIfClosed();
        if (cal == null || cal.countedRefHandle == 0L) {
            throw new IllegalArgumentException("rebin requires a counted-ref EnergyCalibration"
                + " (call toCountedRef() on it first)");
        }
        if (!Native.measurementRebin(handle, cal.countedRefHandle)) {
            throw new SpecUtilsException("rebin failed");
        }
    }

    public void setEnergyCalibration(EnergyCalibration cal) {
        throwIfClosed();
        if (cal == null || cal.countedRefHandle == 0L) {
            throw new IllegalArgumentException("setEnergyCalibration requires a counted-ref"
                + " EnergyCalibration (call toCountedRef() on it first)");
        }
        if (!Native.measurementSetEnergyCalibration(handle, cal.countedRefHandle)) {
            throw new SpecUtilsException("measurement setEnergyCalibration failed"
                + " (wrong channel count or invalid cal)");
        }
    }

    /**
     * Non-owning access to this Measurement's energy calibration. Returns
     * null if none is set. The returned wrapper's {@link EnergyCalibration#close}
     * is a no-op — its lifetime follows this Measurement.
     */
    public EnergyCalibration energyCalibration() {
        throwIfClosed();
        long rh = Native.measurementEnergyCalibrationPtr(handle);
        if (rh == 0L) return null;
        return new EnergyCalibration(rh, 0L, false, false);
    }

    /**
     * Returns a counted-ref copy of this measurement's energy calibration,
     * owned by the returned wrapper. Use this to keep the calibration
     * alive independently of the Measurement.
     * @return null if measurement has no calibration.
     */
    public EnergyCalibration energyCalibrationRef() {
        throwIfClosed();
        long crh = Native.measurementEnergyCalibrationRef(handle);
        if (crh == 0L) return null;
        long rh = Native.energyCalPtrFromRef(crh);
        return new EnergyCalibration(rh, crh, false, true);
    }

    // ----- GPS & rates -----

    public boolean hasGpsInfo()    { throwIfClosed(); return Native.measurementHasGpsInfo(handle); }
    public double latitude()       { throwIfClosed(); return Native.measurementLatitude(handle); }
    public double longitude()      { throwIfClosed(); return Native.measurementLongitude(handle); }
    public long positionTimeMicroseconds() {
        throwIfClosed();
        return Native.measurementPositionTimeMicrosec(handle);
    }

    public void setPosition(double longitude, double latitude, long positionTimeMicroseconds) {
        throwIfClosed();
        Native.measurementSetPosition(handle, longitude, latitude, positionTimeMicroseconds);
    }

    public float speed()        { throwIfClosed(); return Native.measurementSpeed(handle); }
    public float doseRate()     { throwIfClosed(); return Native.measurementDoseRate(handle); }
    public float exposureRate() { throwIfClosed(); return Native.measurementExposureRate(handle); }

    // ----- Status enums -----

    public OccupancyStatus occupancyStatus() {
        throwIfClosed();
        return OccupancyStatus.fromOrdinal(Native.measurementOccupancyStatus(handle));
    }

    public void setOccupancyStatus(OccupancyStatus s) {
        throwIfClosed();
        Native.measurementSetOccupancyStatus(handle, s.ordinal());
    }

    public QualityStatus qualityStatus() {
        throwIfClosed();
        return QualityStatus.fromOrdinal(Native.measurementQualityStatus(handle));
    }

    public SourceType sourceType() {
        throwIfClosed();
        return SourceType.fromOrdinal(Native.measurementSourceType(handle));
    }

    public void setSourceType(SourceType s) {
        throwIfClosed();
        Native.measurementSetSourceType(handle, s.ordinal());
    }

    // ----- Remarks & warnings -----

    public int numberRemarks() { throwIfClosed(); return Native.measurementNumberRemarks(handle); }
    public String remark(int index) { throwIfClosed(); return Native.measurementRemark(handle, index); }
    public void setRemarks(String[] remarks) {
        throwIfClosed();
        Native.measurementSetRemarks(handle, remarks);
    }

    public int numberParseWarnings() {
        throwIfClosed();
        return Native.measurementNumberParseWarnings(handle);
    }
    public String parseWarning(int index) {
        throwIfClosed();
        return Native.measurementParseWarning(handle, index);
    }

    @Override
    public synchronized void close() {
        if (handle == 0L) return;
        if (cleanerRegistration != null) {
            cleanerRegistration.deregister();
            cleanerRegistration = null;
        }
        if (refHandle != 0L) {
            // Measurement obtained from a SpecFile: release our reference-counted handle (the
            // underlying measurement is freed once all references and the SpecFile are gone).
            Native.countedRefMeasurementDestroy(refHandle);
        } else if (ownsHandle) {
            Native.measurementDestroy(handle);
        }
        handle = 0L;
        refHandle = 0L;
        ownsHandle = false;
    }

    private void throwIfClosed() {
        if (handle == 0L) {
            throw new IllegalStateException("Measurement has been closed");
        }
    }
}
